### PR TITLE
Applied email preference when staff emails students in bulk

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
#### What are the relevant tickets?
- fixes https://github.com/mitodl/micromasters/issues/1781

#### What's this PR do?
- Applied email preference (profile. email_optin) on emails recipients, when staff emails students in bulk

@pdpinch @noisecapella @giocalitri

### CODE REVIEW

File: dashboard/serializers.py [hunk 1]
```diff
@@ -67,3 +69,14 @@ def is_learner(cls, user, program):
...
+
+    @classmethod
+    def email_optin(cls, user):
+        """
+        Returns true if email_optin is set true or false
+
+        Args:
+            user (django.contrib.auth.models.User): A user
+        """
+        user_profile = Profile.objects.get(user=user)
```
Comment (noisecapella) [inline dashboard/serializers.py]: Why not just `user.profile` here?

File: dashboard/serializers.py [hunk 2]
```diff
@@ -67,3 +69,14 @@ def is_learner(cls, user, program):
         return (
             not Role.objects.filter(user=user, role__in=NON_LEARNERS, program=program).exists()
         )
+
+    @classmethod
+    def email_optin(cls, user):
+        """
+        Returns true if email_optin is set true or false
```
Comment (noisecapella) [inline dashboard/serializers.py]: Can you remove `true or false` at the end of this line?

File: dashboard/serializers.py [hunk 3]
```diff
@@ -67,3 +69,14 @@ def is_learner(cls, user, program):
...
+
+        Args:
```
Comment (noisecapella) [inline dashboard/serializers.py]: Can you add a `Returns:` declaration here?

File: dashboard/serializers_test.py [hunk 4]
```diff
@@ -129,14 +130,30 @@ def test_full_program_user_serialization(self):
...
             'enrollments': list(CachedEnrollment.active_data(self.user, program)),
             'certificates': list(CachedCertificate.active_data(self.user, program)),
             'current_grades': list(CachedCurrentGrade.active_data(self.user, program)),
             'grade_average': 75,
-            'is_learner': True
+            'is_learner': True,
+            'email_optin': False
+        }
+
+        Profile.objects.filter(pk=self.profile.pk).update(email_optin=True)
```
Comment (noisecapella) [inline dashboard/serializers_test.py]: Can you make this a separate test?

File: dashboard/serializers_test.py [hunk 5]
```diff
@@ -173,7 +208,21 @@ def test_full_program_user_serialization_staff(self):
             'certificates': list(CachedCertificate.active_data(self.user, program)),
             'current_grades': list(CachedCurrentGrade.active_data(self.user, program)),
             'grade_average': 75,
-            'is_learner': False
+            'is_learner': False,
+            'email_optin': False
+        }
+
+        Profile.objects.filter(pk=self.profile.pk).update(email_optin=True)
```
Comment (noisecapella) [inline dashboard/serializers_test.py]: Can you make this a separate test?

File: dashboard/serializers_test.py [hunk 6]
```diff
@@ -146,20 +163,38 @@ def test_full_program_user_serialization_financial_aid(self):
...
             'certificates': [],
             'current_grades': list(CachedCurrentGrade.active_data(self.user, self.fa_program)),
             'grade_average': 95,
-            'is_learner': True
+            'is_learner': True,
+            'email_optin': False
+        }
+        assert UserProgramSearchSerializer.serialize(self.fa_program_enrollment) == expected_result
+
+        Profile.objects.filter(pk=self.profile.pk).update(email_optin=True)
```
Comment (noisecapella) [inline dashboard/serializers_test.py]: Make this a separate test please.

Comment (noisecapella) [inline dashboard/serializers_test.py]: Actually I don't think we need to test `email_optin` here or below. We only need to test it in one place, it's not linked to any other field in the profile.

File: search/api.py [hunk 7]
```diff
@@ -29,7 +29,7 @@ def execute_search(search_obj):
     return search_obj.execute()
 
 
-def create_program_limit_query(user):
+def create_program_limit_query(user, omit_email_optin_false=False):
```
Comment (noisecapella) [inline search/api.py]: Please rename to `omit_email_optin`, and please add information on the arguments and return value to the docstring

File: search/api.py [hunk 8]
```diff
@@ -50,13 +57,11 @@ def create_program_limit_query(user):
...
         minimum_should_match=1,
-        must=[
-            Q('term', **{'program.is_learner': True})
-        ],
+        must=must,
     )
 
 
-def create_search_obj(user, search_param_dict=None):
+def create_search_obj(user, search_param_dict=None, omit_email_optin_false=False):
```
Comment (noisecapella) [inline search/api.py]: Please rename to `filter_on_email_optin` and add this to the docstring

File: search/api.py [hunk 9]
```diff
@@ -80,11 +88,15 @@ def create_search_obj(user, search_param_dict=None):
     return search_obj
 
 
-def prepare_and_execute_search(user, search_param_dict=None, search_func=execute_search):
+def prepare_and_execute_search(user, search_param_dict=None, search_func=execute_search, omit_email_optin_false=False):
```
Comment (noisecapella) [inline search/api.py]: Please rename to `filter_on_email_optin`

File: search/api_test.py [hunk 10]
```diff
@@ -174,6 +183,18 @@ def test_search_user(self):
         self.assertTrue(results[0].program.is_learner)
         self.assertTrue(results[1].program.is_learner)
 
+        # when omit_email_optin_false is enable.
```
Comment (noisecapella) [inline search/api_test.py]: Can you make this a separate test?

File: search/api_test.py [hunk 11]
```diff
@@ -174,6 +183,18 @@ def test_search_user(self):
...
+            search_param_dict=params,
+            omit_email_optin_false=True
+        )
+
+        self.assertEqual(len(results), 1)
+        self.assertEqual(results[0].user_id, learner.id)
+        self.assertTrue(results[0].program.is_learner)
+        self.assertTrue(results[0].program.email_optin)
+
     # def test_all_query_matching_results(self):
```
Comment (noisecapella) [inline search/api_test.py]: Would you mind removing this line while you're editing this file?

Comment (amir-qayyum-khan) [inline search/api_test.py]: sure

**Comment by noisecapella** [comment]
Can you edit the `test_send_view` test in `mail/views_test.py` to assert the arguments passed to `prepare_exec_search`? https://github.com/mitodl/micromasters/blob/master/mail/views_test.py#L72

**Comment by amir-qayyum-khan** [comment]
Done @noisecapella

File: dashboard/serializers_test.py (hunk 4 above)
Comment (noisecapella) [inline dashboard/serializers_test.py]: Please make a separate test here to separate the `email_optin` true and false cases. You can use the `ddt` library to test true and false. See here for an example: https://github.com/mitodl/micromasters/blob/master/financialaid/api_test.py#L162

**Comment by noisecapella** [comment]
I made this comment above: https://github.com/mitodl/micromasters/pull/1842#discussion_r89362120

**Comment by amir-qayyum-khan** [comment]
done @noisecapella

**Pull request merged**